src/server/mod.rs

```diff
@@ -82,6 +83,12 @@ pub struct Builder<I> {
     protocol: Http_,
 }
 
+/// AsyncRewind is a rewindable AsyncRead+AsyncWrite trait
+pub trait AsyncRewind: AsyncRead + AsyncWrite {
```

Comment:
I think we can probably build the rewinding in internal, and not require that ever transport type a user might want to use have to implement this function. Basically, a private `struct Rewound<I>`. Here's a way I'd do it:

1. In the h1 case, nothing should be needed there. If there is a parse error, we can just use `deconstruct`/`into_inner` to get the original transport and the `read_buf` back.
2. If we've done number 1, and want to start the h2 handshake, then a `Rewound` would be constructed with those 2 pieces, and then `Rewound<I>` would *be* the `AsyncRead + AsyncWrite` type passed to the h2 handshake/connection.

That'd make the `server::Connection` type look kind of like this internally:

```rust
pub struct Connection<I, S> {
    inner: Option<
        Either<
            h1::Dispatcher<..., I>,
            h2::Server<Rewound<I>, ...>,
        >
    >
}
```